lib/segment/src/vector_storage/async_io.rs

```diff
@@ -73,72 +71,240 @@ impl UringBufferedReader {
-                callback(idx, point, vector);
-                unused_buffer_ids.push(buffer_id);
+        // Wait for at least one CQE.
+        probe_or_wait_for_cqe(&submitter, &mut cq)?;
+
+        // TODO
+        while submitted > 0 {
+            // TODO
+            for cqe in &mut cq {
+                // Track consumed CQE.
+                submitted -= 1;
+
+                // Process consumed CQE.
+                let buffer_index = cqe.user_data() as usize;
+                let buffer = &mut self.buffers[buffer_index];
+
+                let point = buffer
+                    .point
+                    .take()
+                    .expect("point data is associated with the buffer");
+                let vector = transmute_from_u8_to_slice(&buffer.buffer);
+
+                callback(point.index, point.point, &vector);
+
+                // Check if there are more SQEs to push and submit.
+                let Some((point_index, point)) = points.next() else {
+                    if sq.is_empty() {
+                        // All points exhausted and SQ is empty.
+                        // Continue processing CQEs in the inner loop until CQ is empty.
+                        continue;
+                    } else {
+                        // All points exhausted and SQ is not empty.
+                        // Break out of the inner loop to submit *the last* SQEs.
+                        break;
+                    }
+                };
+
+                // TODO: Deduplicate with similar snippet in SQ loop?
+                let point_offset_bytes =
+                    self.header_size_bytes + self.vector_size_bytes * point as usize;
+
+                let sqe = io_uring::opcode::Read::new(
+                    types::Fd(self.file.as_raw_fd()),
+                    buffer.buffer.as_mut_ptr(),
+                    buffer.buffer.len() as _,
+                )
+                .offset(point_offset_bytes as _)
+                .build()
+                .user_data(buffer_index as _);
+
+                buffer.point = Some(Point::new(point_index, point));
+
+                unsafe { sq.push(&sqe).expect("SQ is not full") };
+
+                if sq.is_full() {
+                    // SQ is full.
+                    // Break out of the inner loop to submit *more* SQEs.
+                    break;
+                }
             }
 
-            operations_to_wait_for = self.buffers.buffers.len() - unused_buffer_ids.len();
+            if sq.is_empty() {
+                probe_or_wait_for_cqe(&submitter, &mut cq)?;
+            } else {
+                submitted += submit_sqes_and_wait_for_cqe(&submitter, &mut sq, &mut cq)?;
+            }
         }
+
         Ok(())
     }
 }
+
+fn check_cq_is_empty(cq: &mut io_uring::CompletionQueue<'_>) -> bool {
+    cq.sync();
+    cq.is_empty()
+}
+
+fn probe_cq_is_empty(cq: &mut io_uring::CompletionQueue<'_>) -> bool {
+    for _ in 0..3 {
```

Comment:
?

Reply:
This probes if CQ is empty with `sync` calls (i.e., atomic operations) three times. The reasoning is that atomic ops should be much faster than a call to `submit_and_wait`, so we try to busy wait a tiny bit before comitting to a `submit_and_wait` call.